File: ai-agent/app/schemas/notion_agent.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.rag import AgentStep


class NotionSourceChunk(BaseModel):
    page_id: str
    page_title: str
    url: str
    chunk_index: int
    text: str
    score: float


class NotionSourcePage(BaseModel):
    page_id: str
    page_title: str
    url: str
    chunk_count: int
    last_edited_time: datetime | None = None
    indexed_at: datetime


class NotionSyncResponse(BaseModel):
    indexed_pages: int
    indexed_chunks: int
    skipped_pages: int
    message: str


class ListNotionSourcesResponse(BaseModel):
    sources: list[NotionSourcePage]


class NotionAgentChatRequest(BaseModel):
    session_id: str | None = Field(default=None, min_length=1, max_length=128)
    thread_id: str | None = Field(default=None, min_length=1, max_length=128)
    client_id: str | None = Field(default=None, min_length=1, max_length=128)
    message: str = Field(..., min_length=1)
    top_k: int | None = Field(default=None, ge=1, le=20)


class NotionAgentChatResponse(BaseModel):
    answer: str
    mode: Literal["notion_rag"] = "notion_rag"
    sources: list[NotionSourceChunk] = Field(default_factory=list)
    agent_steps: list[AgentStep] = Field(default_factory=list)
    session_id: str
    thread_id: str

File: client/src/api/notionAgent.ts
import axios from "axios";
import type { AgentStep } from "./rag";

const AI_AGENT_BASE_URL = import.meta.env.VITE_AI_AGENT_API_URL || "http://localhost:8001";

const notionAgentApi = axios.create({
  baseURL: AI_AGENT_BASE_URL,
});

export type NotionSourceChunk = {
  page_id: string;
  page_title: string;
  url: string;
  chunk_index: number;
  text: string;
  score: number;
};

export type NotionSourcePage = {
  page_id: string;
  page_title: string;
  url: string;
  chunk_count: number;
  last_edited_time: string | null;
  indexed_at: string;
};

export type NotionSyncResponse = {
  indexed_pages: number;
  indexed_chunks: number;
  skipped_pages: number;
  message: string;
};

export type NotionAgentChatResponse = {
  answer: string;
  mode: "notion_rag";
  sources: NotionSourceChunk[];
  agent_steps: AgentStep[];
  session_id: string;
  thread_id: string;
};

export const syncNotionMemory = async (): Promise<NotionSyncResponse> => {
  const response = await notionAgentApi.post<NotionSyncResponse>("/notion-agent/sync");
  return response.data;
};

export const getNotionSources = async (): Promise<NotionSourcePage[]> => {
  const response = await notionAgentApi.get<{ sources: NotionSourcePage[] }>(
    "/notion-agent/sources",
  );
  return response.data.sources;
};

export const sendNotionAgentMessage = async (
  threadId: string,
  clientId: string,
  message: string,
): Promise<NotionAgentChatResponse> => {
  const response = await notionAgentApi.post<NotionAgentChatResponse>(
    "/notion-agent/chat",
    {
      thread_id: threadId,
      client_id: clientId,
      message,
    },
  );
  return response.data;
};
